- Add modularized OpenPlanner to autoware , 2nd PR

hatem-darweesh opened the pull request:
## Status
**DEVELOPMENT**

## Description
- Update OpenPlanner libraries (op_planner, op_utility, op_simu)
- Add new library for ROS helper functions (op_ros_helpers)
- Add Nodes for local planning 
1) op_global_planner
2) op_common_params
3) op_motion_predictor
4) op_trajectory_evaluator
5) op_behavior_selector

- Add Simulation Nodes (op_simulation_package)
1) op_car_simulator
2) op_perception_simulator
3) op_signs_simulator

- Add UI launchers in runtime manager (computing tab)

## Steps to Test or Reproduce
```
From the Runtime manager run the OpenPlanner nodes as illustrated in the video: 
https://youtu.be/BS5nLtBsXPE
```
results should be as the results in the video.

- To do: 
Provide design/testing documents for all functionality in the library and ROS nodes.

hatem-darweesh committed:
- Add modularized OpenPlanner to autoware , 2nd PR
1) op_global_planner
2) op_common_params
3) op_motion_predictor
4) op_trajectory_evaluator
5) op_behavior_selector

- Add Simulation Nodes (op_simulation_package)
1) op_car_simulator
2) op_perception_simulator
3) op_signs_simulator

- Add UI launchers in runtime manager (computing tab)

Capri2014 commented on ros/src/computing/planning/common/lib/openplanner/op_planner/include/DecisionMaker.h:
```diff
@@ -0,0 +1,91 @@
+/*
+ * CarState.h
```
Should this be "DeceisionMaker.h"?

hatem-darweesh replied:
fixed

Capri2014 commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/DecisionMaker.cpp:
```diff
@@ -0,0 +1,503 @@
+/*
+ * CarState.cpp
```
Same here

hatem-darweesh replied:
fixed

Capri2014 commented on ros/src/computing/planning/motion/packages/op_local_planner/nodes/op_behavior_selector/op_behavior_selector_core.cpp:
```diff
@@ -0,0 +1,495 @@
+void BehaviorGen::UpdatePlanningParams(ros::NodeHandle& _nh)
+{
+	_nh.getParam("/op_common_params/enableSwerving", m_PlanningParams.enableSwerving);
+	if(m_PlanningParams.enableSwerving)
```
If I understand it right, if "enableSwerving" is set, "enableFollowing" would be set and then in "StateMachine" the state would be updated to "Following" or "Obstacle Avoidance" depending on whether obstacle is fully blocking or not. Would it be more reasonable to incorporate this part of logic into StateMachine instead of doing it in different places ? Like  "BehaviorGen", "TrajectoryGen", "MotionPrediction", "TrajectoryEval" separately?

hatem-darweesh replied:
It is implemented as state machine in behavior_selector , the condition just control if certain states are included in the loop or not, and because "following" and "avoidance" is related conditionally - following could be used without avoidance, but when avoidance is activated following must be activated by default-

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/CMakeLists.txt:
```diff
@@ -43,6 +43,8 @@ set(PLANNERH_SRC
 		src/PlanningHelpers.cpp
 		src/TrajectoryPrediction.cpp
+		src/BehaviorPrediction.cpp
+		src/DecisionMaker.cpp
```
does this DecisionMaker have anything to do with https://github.com/CPFL/Autoware/tree/develop/ros/src/computing/planning/decision/packages/decision_maker?

hatem-darweesh replied:
I think similar objective but different implementation, 
also I have state machine in behavior selector , and there is another state machine is being implemented in autoware

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/include/BehaviorPrediction.h:
```diff
@@ -0,0 +1,42 @@
+ * BehaviorPrediction.h
+ *
+ *  Created on: Jul 6, 2017
+ *      Author: user
```
this should have an actual name of author

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/include/BehaviorPrediction.h:
```diff
@@ -0,0 +1,42 @@
+/*
+ * BehaviorPrediction.h
```
if you use this syntax:
```
/// \file
/// \brief Line about this file
```

then doxygen will automatically extract file name.

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/include/BehaviorPrediction.h:
```diff
@@ -0,0 +1,42 @@
+public:
+	BehaviorPrediction();
+	virtual ~BehaviorPrediction();
+	void DoOneStep(const std::vector<DetectedObject>& obj_list, const WayPoint& currPose, const double& minSpeed, const double& maxDeceleration, RoadNetwork& map);
```
Comment functions. I highly suggest this syntax:
```
/// \brief one liner
/// \param[in] obj_list ...
/// \param[out] currPose ...
/// \return what should function return
```

dejanpan replied:
obviously all functions in the PR should be commented

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/include/BehaviorStateMachine.h:
```diff
@@ -63,6 +69,15 @@ class ForwardState : public BehaviorStateMachine
+class ForwardStateII : public BehaviorStateMachine
+{
+public:
+	ForwardStateII(PlanningParams* pParams, PreCalculatedConditions* pPreCalcVal, BehaviorStateMachine* pNextState)
```
how do you decide between passing args as pointers or as references?

hatem-darweesh replied:
it depends on the main theme of the data structure.

dejanpan replied:
not the best rule, try to stick to something proven: https://google.github.io/styleguide/cppguide.html#Function_Parameter_Ordering

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/include/PlanningHelpers.h:
```diff
@@ -101,13 +101,17 @@ class PlanningHelpers
 	//static double CalcAngleAndCostSimple(std::vector<WayPoint>& path, const double& lastCost = 0);
 	static double CalcAngleAndCostAndCurvatureAnd2D(std::vector<WayPoint>& path, const double& lastCost = 0);
 
+	static void PredictConstantTimeCostForTrajectory(std::vector<PlannerHNS::WayPoint>& path, const PlannerHNS::WayPoint& currPose, const double& minVelocity, const double& minDist);
```
to make the reviews easier it would be actually great if you guys would use 120 char line width

hatem-darweesh replied:
sorry, it is more readable for me.

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/include/RoadNetwork.h:
```diff
@@ -815,17 +817,18 @@ class PlanningParams
 		horizontalSafetyDistancel		= 0.0;
 
+		nReliableCount					= 2;
+
```
are these params also settable through `runtime_manager` or any other config possibility?

hatem-darweesh replied:
yes, you can change these parameter from runtime manager.

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/BehaviorPrediction.cpp:
```diff
@@ -0,0 +1,132 @@
+	m_MaxLaneDetectionDistance = 0.5;
+	m_PredictionDistance = 20.0;
+	m_bGenerateBranches = false;
+	m_bUseFixedPrediction = true;
```
parametrize

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/BehaviorPrediction.cpp:
```diff
@@ -0,0 +1,132 @@
+	for(unsigned int i=0; i < obj_list.size(); i++)
+	{
+		if(obj_list.at(i).t == SIDEWALK || obj_list.at(i).center.v < 1.0)
+			continue;
```
it is really a good practice to use {} after every control structure

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/BehaviorPrediction.cpp:
```diff
@@ -0,0 +1,132 @@
+			continue;
+
+		bool bFound = false;
+		int found_index = 0;
```
I wonder what is your style guide. In above 2 lines you use 2 styles

hatem-darweesh replied:
too many programming platforms and languages, it is so hard for me to be consistent.

dejanpan replied:
well here it is C++ only. not sure what do you mean by programming platforms.

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/BehaviorPrediction.cpp:
```diff
@@ -0,0 +1,132 @@
+		if(bFound)
+			filtered_list.at(found_index) = obj_list.at(i);
+		else
+			filtered_list.push_back(obj_list.at(i));
```
just something to be aware of. `push_back`, `pop`, ... allocate and deallocate memory which leads to memory fragmentation. It is better to use reserve() to reserve and re-use a contiguous memory.

hatem-darweesh replied:
Thanks, good idea. I will try to use it.

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/BehaviorPrediction.cpp:
```diff
@@ -0,0 +1,132 @@
+void BehaviorPrediction::DoOneStep(const std::vector<DetectedObject>& obj_list, const WayPoint& currPose, const double& minSpeed, const double& maxDeceleration, RoadNetwork& map)
+{
+	if(!m_bUseFixedPrediction)
+		m_PredictionDistance = -pow(currPose.v, 2)/(maxDeceleration);
```
possible division by 0

hatem-darweesh replied:
done

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/BehaviorStateMachine.cpp:
```diff
@@ -50,13 +49,54 @@ void BehaviorStateMachine::InsertNextState(BehaviorStateMachine* nextState)
+		}
+	}
+
+	return 0;
```
you should return `nullptr` if you do not end up in above if

hatem-darweesh replied:
done

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/BehaviorStateMachine.cpp:
```diff
@@ -86,10 +126,12 @@ BehaviorStateMachine* ForwardState::GetNextState()
 
-	else if(m_pParams->enableSwerving
+//	if(pCParams->currentGoalID != pCParams->prevGoalID)
+//		return FindBehaviorState(GOAL_STATE);
```
remove dead code

hatem-darweesh replied:
"may the dead never die" , it may make a come back one day. I try to keep it minimal , I know it looks ugly.

dejanpan replied:
It is not just about the looks, it has practical implications:
- code is less readable
- if someone else looks through the code and tries to find a bug, they might start uncommenting the dead code but they do not have a context
- code will become obsolete

Use git history.

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/BehaviorStateMachine.cpp:
```diff
@@ -305,4 +338,123 @@ BehaviorStateMachine* GoalState::GetNextState()
+
+BehaviorStateMachine* SwerveStateII::GetNextState()
+{
+	PreCalculatedConditions* pCParams = GetCalcParams();
```
in general you should check if the pointer is valid before you try to use it

hatem-darweesh replied:
sure, but GetCalcParams() never return an invalid point.

dejanpan replied:
sure it can. Like if someone has by mistake manipulated the memory address of GetCalcParams. Or if the function was not able allocate the memory because of the mem fragmentation issue. Etc.

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/DecisionMaker.cpp:
```diff
@@ -0,0 +1,503 @@
+	m_pGoToGoalState 		=0;
+	m_pWaitState 			=0;
+	m_pInitState 			=0;
+	m_pFollowState			=0;
```
why this alignment?

hatem-darweesh replied:
maybe I had some spare time!

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/DecisionMaker.cpp:
```diff
@@ -0,0 +1,503 @@
+//	m_pTrafficLightStopState	= new TrafficLightStopState(m_pStopState->m_pParams, m_pStopState->GetCalcParams(), m_pGoToGoalState);
+//	m_pTrafficLightWaitState	= new TrafficLightWaitState(m_pStopState->m_pParams, m_pStopState->GetCalcParams(), m_pGoToGoalState);
+	m_pStopSignWaitState		= new StopSignWaitStateII(m_pStopState->m_pParams, m_pStopState->GetCalcParams(), m_pGoToGoalState);
+	m_pStopSignStopState		= new StopSignStopStateII(m_pStopState->m_pParams, m_pStopState->GetCalcParams(), m_pStopSignWaitState);
```
Did you consider using smart_ptr too?

The rule would be this - if you know that an entity must take a certain kind of ownership of the object, always use smart pointers - the one that gives you the kind of ownership you need. If there is no notion of ownership, never use smart pointers.

hatem-darweesh replied:
will consider it , but as long as I can achieve the same objective using classical c pointer I go for it.

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/DecisionMaker.cpp:
```diff
@@ -0,0 +1,503 @@
+//	m_pTrafficLightStopState->InsertNextState(m_pTrafficLightWaitState);
+//	m_pTrafficLightWaitState->InsertNextState(m_pTrafficLightStopState);
+//
+	m_pStopSignWaitState->decisionMakingTime = 5.0;
```
parametrize

hatem-darweesh replied:
This functionality is not complete yet, will add to the parameterization plan

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/DecisionMaker.cpp:
```diff
@@ -0,0 +1,503 @@
+
+ 	PreCalculatedConditions* pValues = m_pCurrentBehaviorState->GetCalcParams();
+
+ 	pValues->minStoppingDistance = -pow(car_state.speed, 2)/(m_CarInfo.max_deceleration);
```
div by 0

hatem-darweesh replied:
done

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/DecisionMaker.cpp:
```diff
@@ -0,0 +1,503 @@
+	double critical_long_front_distance = m_CarInfo.length/2.0;
+
+
+	if(m_Path.size() <= 5)
```
parametrize

hatem-darweesh replied:
will think about better solution than this, added to the changing plan.

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/DecisionMaker.cpp:
```diff
@@ -0,0 +1,503 @@
+
+			if(desiredVelocity > max_velocity)
+				desiredVelocity = max_velocity;
+			else if(desiredVelocity < 0.25)
```
parametrize

hatem-darweesh replied:
will add it to the plan

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/DecisionMaker.cpp:
```diff
@@ -0,0 +1,503 @@
+//
+//			std::cout << "Follow Dec: Target V: " << desiredVelocity << ", Object D: " <<  follow_d << ", E: " << e << std::endl;
+//			return desiredVelocity;
+//		}
```
remove dead code

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/PlanningHelpers.cpp:
```diff
@@ -1223,7 +1189,7 @@ void PlanningHelpers::ExtractPartFromPointToDistance(const vector<WayPoint>& ori
 	//vector<WayPoint> tempPath;
 	double d_limit = 0;
-	if(close_index >= 5) close_index -=5;
+	if(close_index >= 2) close_index -=2;
```
parametrize

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/TrajectoryCosts.cpp:
```diff
@@ -15,18 +15,202 @@ namespace PlannerHNS
+
+	int smallestIndex = -1;
+	double smallestCost = 9999999999;
+	double smallestDistance = 9999999999;
```
DBL_MAX?

hatem-darweesh replied:
done

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/TrajectoryCosts.cpp:
```diff
@@ -127,6 +311,129 @@ TrajectoryCost TrajectoryCosts::DoOneStep(const vector<vector<vector<WayPoint> >
+					longitudinalDist = -longitudinalDist;
+
+				double direct_distance = hypot(obj_info.perp_point.pos.y-contourPoints.at(icon).pos.y, obj_info.perp_point.pos.x-contourPoints.at(icon).pos.x);
+				if(contourPoints.at(icon).v < 0.1 && direct_distance > (m_LateralSkipDistance+contourPoints.at(icon).cost))
```
param for 0.1

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/TrajectoryCosts.cpp:
```diff
@@ -127,6 +311,129 @@ TrajectoryCost TrajectoryCosts::DoOneStep(const vector<vector<vector<WayPoint> >
+
+
+				trajectoryCosts.at(iCostIndex).lateral_cost += 1.0/lateralDist;
+				trajectoryCosts.at(iCostIndex).longitudinal_cost += 1.0/fabs(longitudinalDist);
```
div by 0

hatem-darweesh replied:
done

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/TrajectoryCosts.cpp:
```diff
@@ -386,96 +682,225 @@ bool TrajectoryCosts::ValidateRollOutsInput(const vector<vector<vector<WayPoint>
 			for(unsigned int i = 0; i < path.size(); i++)
 			{
-				if(path.at(i).timeCost > 0.0)
+				//if(path.at(i).timeCost > -1)
```
remove { below?

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_simu/src/PlannerTestDraw.cpp:
```diff
@@ -1519,8 +1519,8 @@ void* PlannerTestDraw::PlanningThreadStaticEntryPoint(void* pThis)
-			if(pR->m_LocalPlanner.m_pCurrentBehaviorState->GetCalcParams()->bOutsideControl == 1 && pR->m_CurrentBehavior.state != PlannerHNS::INITIAL_STATE)
-				bEmergencyStop = true;
+//			if(pR->m_LocalPlanner.m_pCurrentBehaviorState->GetCalcParams()->bOutsideControl == 1 && pR->m_CurrentBehavior.state != PlannerHNS::INITIAL_STATE)
+//				bEmergencyStop = true;
```
remove dead code

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_ros_helpers/src/op_RosHelpers.cpp:
```diff
@@ -238,73 +238,541 @@ void RosHelpers::CreateCircleMarker(const PlannerHNS::WayPoint& _center, const d
+
+	obj.velocity.linear.x = det_obj.center.v;
+	obj.velocity_reliable = det_obj.bVelocity;
+	obj.pose_reliable = det_obj.bDirection;
```
direction is not pose

hatem-darweesh replied:
I temporary use it this way between OpenPlanner nodes. I should add new variable to autoware_msg. will add a memo

dejanpan commented on ros/src/computing/planning/common/lib/openplanner/op_utility/src/DataRW.cpp:
```diff
@@ -32,6 +34,47 @@ DataRW::~DataRW()
+	dir_err = mkdir(main_folder.c_str(), S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);
+
+	main_folder = UtilityH::GetHomeDirectory() + DataRW::LoggingMainfolderName + DataRW::SimulationFolderName;
+	dir_err = mkdir(main_folder.c_str(), S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);
```
2 blocks above are identical

hatem-darweesh replied:
done

dejanpan commented on ros/src/computing/planning/mission/packages/op_global_planner/nodes/op_global_planner.cpp:
```diff
@@ -0,0 +1,49 @@
+ */
+
+#include <ros/ros.h>
+#include "op_global_planner_core.h"
```
swap, see https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

dejanpan commented on ros/src/computing/planning/mission/packages/op_global_planner/nodes/op_global_planner_core.cpp:
```diff
@@ -0,0 +1,392 @@
+	else if(bVelSource == 1)
+		sub_current_velocity 	= nh.subscribe("/current_velocity",		100,	&GlobalPlanner::callbackGetVehicleStatus, 	this);
+	else if(bVelSource == 2)
+		sub_can_info 			= nh.subscribe("/can_info",		100,	&GlobalPlanner::callbackGetCanInfo, 	this);
```
fomatting of the code is messed up

dejanpan replied:
Otherwise also not the best practice to hard code topics

hatem-darweesh replied:
can you share link with the best practice of topic definition

dejanpan replied:
no magic. set a param in launch file and then read this param in above program. you can still keep the default value.

dejanpan commented on ros/src/computing/planning/mission/packages/op_global_planner/nodes/op_global_planner_core.cpp:
```diff
@@ -0,0 +1,392 @@
+{
+	m_VehicleState.speed = msg->twist.twist.linear.x;
+	m_CurrentPose.v = m_VehicleState.speed;
+	m_VehicleState.steer += atan(2.7 * msg->twist.twist.angular.z/msg->twist.twist.linear.x);
```
parametrize

dejanpan replied:
and possible div by 0

dejanpan commented on ros/src/computing/planning/mission/packages/op_global_planner/nodes/op_global_planner_core.cpp:
```diff
@@ -0,0 +1,392 @@
+	m_CurrentPose.v = m_VehicleState.speed;
+	m_VehicleState.steer = msg->angle * 0.45 / 660;
+	UtilityHNS::UtilityH::GetTickCount(m_VehicleState.tStamp);
+}
```
parametrize here and in the whole file

dejanpan commented on ros/src/computing/planning/mission/packages/op_global_planner/nodes/op_global_planner_core.cpp:
```diff
@@ -0,0 +1,392 @@
+
+	ret = m_PlannerH.PlanUsingDP(startPoint, goalPoint, MAX_GLOBAL_PLAN_DISTANCE, m_params.bEnableLaneChange, predefinedLanesIds, m_Map, generatedTotalPaths);
+
+	if(ret == 0)
```
ret is double, you can not compare for equality. Use:
```
// bool cmpf(float A, float B, float epsilon = 0.005f)
      //{
      //return (fabs(A - B) < epsilon);
      //}
```

hatem-darweesh replied:
here 0 means no path is generated and it is dead zero by design.

dejanpan replied:
I think you did not understand my point. you can not compare floating point numbers, please see: http://floating-point-gui.de/errors/comparison/.

dejanpan commented on ros/src/computing/planning/mission/packages/op_global_planner/nodes/op_global_planner_core.cpp:
```diff
@@ -0,0 +1,392 @@
+
+void GlobalPlanner::MainLoop()
+{
+	ros::Rate loop_rate(10);
```
parametrize 10

dejanpan commented on ros/src/computing/planning/mission/packages/op_global_planner/nodes/op_global_planner_core.cpp:
```diff
@@ -0,0 +1,392 @@
+		}
+		else if (m_params.mapSource == PlannerHNS::MAP_FOLDER && !m_bKmlMap)
+		{
+			m_bKmlMap = true;
```
I don't get this logic

hatem-darweesh replied:
Load the specific map type once  in the begging of the node life time

dejanpan reviewed (changes requested):
I came to op_behavior_selector_core.h file. Will try to finish in the next days but I can not promise.
In general the code is good but there is too many hardcoded things and in general you should check for return values of functions more.

Then of course almost no functions are documented and there is not tests.

dejanpan commented on ros/src/computing/planning/motion/packages/op_local_planner/include/op_behavior_selector_core.h:
```diff
@@ -0,0 +1,161 @@
+#include <ros/ros.h>
+#include "PlannerCommonDef.h"
+#include "TrajectoryCosts.h"
+#include "DecisionMaker.h"
```
above 3 should go at the end according to https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

hatem-darweesh replied:
done

dejanpan commented on ros/src/computing/planning/motion/packages/op_local_planner/include/op_behavior_selector_core.h:
```diff
@@ -0,0 +1,161 @@
+
+	ros::Subscriber sub_twist_cmd			;
+	ros::Subscriber sub_twist_raw			;
+	ros::Subscriber sub_ctrl_cmd			;
```
formatting

hatem-darweesh replied:
done

dejanpan commented on ros/src/computing/planning/motion/packages/op_local_planner/include/op_motion_predictor_core.h:
```diff
@@ -0,0 +1,131 @@
+// ROS includes
+#include <ros/ros.h>
+#include "PlannerCommonDef.h"
+#include "BehaviorPrediction.h"
```
order of includes and just keep this in your mind for the entire PR

hatem-darweesh replied:
will do

dejanpan commented on ros/src/computing/planning/motion/packages/op_local_planner/include/op_motion_predictor_core.h:
```diff
@@ -0,0 +1,131 @@
+	void callbackGetCanInfo(const autoware_msgs::CanInfoConstPtr &msg);
+	void callbackGetRobotOdom(const nav_msgs::OdometryConstPtr& msg);
+
+protected:
```
why do you repeat `protected:` twice?

hatem-darweesh replied:
separate ROS related topics and other node functionality

dejanpan replied:
i actually do not understand the answer but the fact is that 2 `protected` are useless

dejanpan commented on ros/src/computing/planning/motion/packages/op_local_planner/nodes/op_behavior_selector/op_behavior_selector_core.cpp:
```diff
@@ -0,0 +1,495 @@
+
+	sub_twist_raw = nh.subscribe("/twist_raw",		1,		&BehaviorGen::callbackGetTwistRaw, 	this);
+	sub_twist_cmd = nh.subscribe("/twist_cmd",		1,		&BehaviorGen::callbackGetTwistCMD, 	this);
+	//sub_ctrl_cmd = nh.subscribe("/ctrl_cmd",		1,		&BehaviorGen::callbackGetCommandCMD, 	this);
```
formatting messed

hatem-darweesh replied:
done

dejanpan commented on ros/src/computing/planning/motion/packages/op_local_planner/nodes/op_behavior_selector/op_behavior_selector_core.cpp:
```diff
@@ -0,0 +1,495 @@
+
+	PlannerHNS::ControllerParams controlParams;
+	controlParams.Steering_Gain = PlannerHNS::PID_CONST(0.07, 0.02, 0.01);
+	controlParams.Velocity_Gain = PlannerHNS::PID_CONST(0.1, 0.005, 0.1);
```
are these param generic and not vehicle specific? I know that they are just initial conditions but still

hatem-darweesh replied:
yes, they should be tuned for every robot, and should be parameterized.

dejanpan commented on ros/src/computing/planning/motion/packages/op_local_planner/nodes/op_behavior_selector/op_behavior_selector_core.cpp:
```diff
@@ -0,0 +1,495 @@
+	_nh.getParam("/op_common_params/mapFileName" 		, m_MapPath);
+
+	_nh.getParam("/op_behavior_selector/evidence_tust_number", m_PlanningParams.nReliableCount);
+	m_PlanningParams.additionalBrakingDistance = 5.0;
```
param

dejanpan commented on ros/src/computing/planning/motion/packages/op_local_planner/nodes/op_behavior_selector/op_behavior_selector_core.cpp:
```diff
@@ -0,0 +1,495 @@
+{
+	m_VehicleStatus.speed = msg->twist.twist.linear.x;
+	m_CurrentPos.v = m_VehicleStatus.speed ;
+	if(msg->twist.twist.linear.x != 0)
```
eq comparison with float/double

hatem-darweesh replied:
just to avoid division by zero

dejanpan replied:
again see my comment above, you can not compare 2 floating point numbers.

dejanpan commented on ros/src/computing/planning/motion/packages/op_local_planner/nodes/op_behavior_selector/op_behavior_selector_core.cpp:
```diff
@@ -0,0 +1,495 @@
+
+void BehaviorGen::MainLoop()
+{
+	ros::Rate loop_rate(100);
```
why 100?  could be a param?

hatem-darweesh replied:
yes it should :) 
but it needs to be Autoware related system param.

dejanpan commented on ros/src/computing/planning/motion/packages/op_local_planner/nodes/op_common_params/op_common_params.cpp:
```diff
@@ -0,0 +1,49 @@
+{
+	ros::init(argc, argv, "op_common_params");
+
+	UtilityHNS::DataRW::CreateLoggingFolder();
```
you should check return value here

hatem-darweesh replied:
of which function ? 
CreateLoggingFolder() doesn't return a value

dejanpan replied:
so what you do if the logging folder couldn't be created?

dejanpan commented on ros/src/computing/planning/motion/packages/op_local_planner/nodes/op_motion_predictor/op_motion_predictor.cpp:
```diff
@@ -0,0 +1,40 @@
+ *  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+#include <ros/ros.h>
+#include "../include/op_motion_predictor_core.h"
```
don't use ..

dejanpan commented on ros/src/computing/planning/motion/packages/op_local_planner/nodes/op_motion_predictor/op_motion_predictor_core.cpp:
```diff
@@ -0,0 +1,310 @@
+	m_VehicleStatus.speed = msg->twist.linear.x;
+	m_CurrentPos.v = m_VehicleStatus.speed;
+	if(fabs(msg->twist.linear.x) > 0.25)
+		m_VehicleStatus.steer = atan(m_CarInfo.wheel_base * msg->twist.angular.z/msg->twist.linear.x);
```
here you guard against division by 0 but what happens then to `m_VehicleStatus.steer`?

hatem-darweesh replied:
zero if first call, last value otherwise

dejanpan commented on ros/src/computing/planning/motion/packages/op_local_planner/nodes/op_motion_predictor/op_motion_predictor_core.cpp:
```diff
@@ -0,0 +1,310 @@
+			if(curb_obstacles.size()>0)
+			{
+				double distance_to_prev = hypot(curb_obstacles.at(curb_obstacles.size()-1).center.pos.y-obj.center.pos.y, curb_obstacles.at(curb_obstacles.size()-1).center.pos.x-obj.center.pos.x);
+				if(distance_to_prev < 1.5)
```
param

dejanpan commented on ros/src/computing/planning/motion/packages/op_local_planner/nodes/op_motion_predictor/op_motion_predictor_core.cpp:
```diff
@@ -0,0 +1,310 @@
+void MotionPrediction::MainLoop()
+{
+
+	ros::Rate loop_rate(25);
```
param

dejanpan commented on ros/src/computing/planning/motion/packages/op_local_planner/nodes/op_trajectory_evaluator/op_trajectory_evaluator_core.cpp:
```diff
@@ -0,0 +1,321 @@
+//
+//				PlannerHNS::PlanningHelpers::GenerateRecommendedSpeed(m_GlobalPaths.at(i), m_CarInfo.max_speed_forward, m_PlanningParams.speedProfileFactor);
+//				m_GlobalPaths.at(i).at(m_GlobalPaths.at(i).size()-1).v = 0;
+//			}
```
remove dead code

dejanpan commented on ros/src/computing/planning/motion/packages/op_simulation_package/CMakeLists.txt:
```diff
@@ -0,0 +1,67 @@
+SET(CMAKE_CXX_FLAGS "-std=c++11 -O2 -g -Wall -Wno-unused-result -DROS ${CMAKE_CXX_FLAGS}")
+
+if ("${ROS_VERSION}" MATCHES "(kinetic)")
+	SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DROS_KINETIC")
```
so the planner really only works in kinetic?

hatem-darweesh replied:
sure not, 
it is the common CMakeList used by Autoware team !

dejanpan replied:
we right now still support both indigo and kinetic.

dejanpan commented on ros/src/computing/planning/motion/packages/op_simulation_package/launch/op_car_simulator_2.launch:
```diff
@@ -0,0 +1,116 @@
+		
+	</node>
+
+</launch>
```
some documentation what all of these params do would be great

hatem-darweesh replied:
I totally agree with you :)

dejanpan commented on ros/src/computing/planning/motion/packages/op_simulation_package/nodes/op_car_simulator/op_car_simulator_core.cpp:
```diff
@@ -0,0 +1,770 @@
+  m1.pose.orientation = tf::createQuaternionMsgFromRollPitchYaw(0, 0, UtilityHNS::UtilityH::SplitPositiveAngle(curr_pose.pos.a));
+  m1.color = m_SimParams.modelColor;
+  m1.scale.x = 1.0*m_CarInfo.length/4.2;
+  m1.scale.y = 1.0*m_CarInfo.width/1.85;
```
param

dejanpan commented on ros/src/msgs/autoware_msgs/msg/waypoint.msg:
```diff
@@ -6,4 +6,23 @@ geometry_msgs/PoseStamped pose
+
+float32 cost
+float32 time_cost
+
```
will this changes be back compatiable? At the very least it will break bags files that you already recorded...

hatem-darweesh replied:
I need to check with other members of the team.

dejanpan reviewed (changes requested):
I went through all of the files but there is simply too many of them to review them in couple of hours.

hatem-darweesh committed:
- Remove commented code
- Adjust some code style issues
- Add files description to OpenPlanner library
- Use nullptr for comparison
- Parameterize hard coded values
- Fix possible division by zero
- Fix synchronization problem
- Fix car following problem

hatem-darweesh committed:
Merge branch 'develop' into featur/OpenPlanner-PR-fix
